injector/injector.go

```diff
@@ -0,0 +1,285 @@
+	if vastXML == "" && NURL == "" {
+		// TODO Log a adapter.<bidder-name>.requests.badserverresponse
+		return vastXML
+	}
+
+	if vastXML == "" {
+		return fmt.Sprintf(emptyAdmResponse, NURL)
+	}
+
+	var outputXML strings.Builder
+	encoder := xml.NewEncoder(&outputXML)
+
+	injectTracker := false
+	injectVideoClicks := false
+	inlineWrapperTagFound := false
+	wrapperTagFound := false
+	impressionTagFound := false
+	errorTagFound := false
+	creativeId := ""
+	isCreative := false
+	companionTagFound := false
+	nonLinearTagFound := false
+
+	reader := strings.NewReader(vastXML)
+	decoder := xml.NewDecoder(reader)
+
+	for {
+		t, err := decoder.RawToken()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return ""
+		}
+
+		switch tt := t.(type) {
+		case xml.StartElement:
+			switch tt.Name.Local {
+			case "Wrapper":
+				wrapperTagFound = true
+			case "Creative":
+				isCreative = true
+				for _, attr := range tt.Attr {
+					if strings.ToLower(attr.Name.Local) == "adid" {
+						creativeId = attr.Value
+					}
+				}
+			case "Linear":
+				injectVideoClicks = true
+				injectTracker = true
+			case "VideoClicks":
+				injectVideoClicks = false
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				ti.addClickTrackingEvent(&outputXML, creativeId)
+
+				continue
+			case "NonLinearAds":
+				injectTracker = true
+			case "TrackingEvents":
+				if isCreative {
+					injectTracker = false
+					encoder.Flush()
+					encoder.EncodeToken(tt)
+					encoder.Flush()
+					ti.addTrackingEvent(&outputXML, creativeId)
+					continue
+				}
+			}
+
+		case xml.EndElement:
+			switch tt.Name.Local {
+			case "Impression":
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				if !impressionTagFound {
+					ti.addImpressionTrackingEvent(&outputXML)
+					impressionTagFound = true
+				}
+				continue
+			case "Error":
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				if !errorTagFound {
+					ti.addErrorTrackingEvent(&outputXML)
+					errorTagFound = true
+				}
+				continue
+			case "NonLinearAds":
+				if injectTracker {
+					injectTracker = false
+					encoder.Flush()
+					outputXML.WriteString("<TrackingEvents>")
+					ti.addTrackingEvent(&outputXML, creativeId)
+					outputXML.WriteString("</TrackingEvents>")
+					if !nonLinearTagFound && wrapperTagFound {
+						outputXML.WriteString("<NonLinear>")
+						ti.addNonLinearClickTrackingEvent(&outputXML, creativeId)
+						outputXML.WriteString("</NonLinear>")
+					}
+					encoder.EncodeToken(tt)
+				}
+			case "Linear":
+				if injectVideoClicks {
+					injectVideoClicks = false
+					encoder.Flush()
+					outputXML.WriteString("<VideoClicks>")
+					ti.addClickTrackingEvent(&outputXML, creativeId)
+					outputXML.WriteString("</VideoClicks>")
+				}
+				if injectTracker {
+					injectTracker = false
+					encoder.Flush()
+					outputXML.WriteString("<TrackingEvents>")
+					ti.addTrackingEvent(&outputXML, creativeId)
```

Comment:
In commit 9677a0e2, the code was modified as follows:
```
167  					injectTracker = false
168  					encoder.Flush()
169  					outputXML.WriteString("<TrackingEvents>")
170 -					for typ, urls := range ti.events.LinearTrackingEvents {
171 -						ti.provider.PopulateEventMacros(creativeId, "", typ)
172 -						for _, url := range urls {
173 -							outputXML.WriteString("<Tracking event=\"")
174 -							outputXML.WriteString(typ)
175 -							outputXML.WriteString("\"><![CDATA[")
176 -							ti.replacer.Replace(&outputXML, url, ti.provider)
177 -							outputXML.WriteString("]]></Tracking>")
178 -						}
179 -					}
    +					ti.addTrackingEvent(&outputXML, creativeId)
180  					outputXML.WriteString("</TrackingEvents>")
```
I noticed that in the removed code line 171, `ti.provider.PopulateEventMacros` is passed an empty string as the second parameter, but inside the `ti.addTrackingEvent(&outputXML, creativeId)` function, `ti.provider.Populate` is called with the string `"tracking"` as the second parameter. Is this discrepancy significant?
 Same happens in the next substitution:
```
206                 if injectTracker {
207                     injectTracker = false
208                     encoder.Flush()
209                     outputXML.WriteString("<TrackingEvents>")
210 -                   for typ, urls := range ti.events.LinearTrackingEvents {
211 -                       ti.provider.PopulateEventMacros(creativeId, "", typ)
212 -                       for _, url := range urls {
213 -                           outputXML.WriteString("<Tracking event=\"")
214 -                           outputXML.WriteString(typ)
215 -                           outputXML.WriteString("\"><![CDATA[")
216 -                           ti.replacer.Replace(&outputXML, url, ti.provider)
217 -                           outputXML.WriteString("]]></Tracking>")
218 -                       }
219 -                   }
    +                   ti.addTrackingEvent(&outputXML, creativeId)
220                     outputXML.WriteString("</TrackingEvents>")
221                 }
```

Reply:
That was discrepancy. Fixed now.